pkg/chunked/storage_linux.go

```diff
@@ -146,40 +146,48 @@ func (c *chunkedDiffer) convertTarToZstdChunked(destDirectory string, payload *o
 func GetDiffer(ctx context.Context, store storage.Store, blobDigest digest.Digest, blobSize int64, annotations map[string]string, iss ImageSourceSeekable) (graphdriver.Differ, error) {
 	pullOptions := store.PullOptions()
 
+	convertImages := parseBooleanPullOption(pullOptions, "convert_images", false)
+
 	if !parseBooleanPullOption(pullOptions, "enable_partial_images", true) {
-		return nil, errors.New("enable_partial_images not configured")
+		return nil, internal.ErrPartialContentNotAvailable
 	}
 
 	zstdChunkedTOCDigestString, hasZstdChunkedTOC := annotations[internal.ManifestChecksumKey]
 	estargzTOCDigestString, hasEstargzTOC := annotations[estargz.TOCJSONDigestAnnotation]
 
 	if hasZstdChunkedTOC && hasEstargzTOC {
-		return nil, errors.New("both zstd:chunked and eStargz TOC found")
+		if !convertImages {
+			return nil, fmt.Errorf("both zstd:chunked and eStargz TOC found: %w", internal.ErrPartialContentNotAvailable)
+		}
+		// If a conversion is required, ignore both TOCs and retrieve the entire layer.
+		hasZstdChunkedTOC, hasEstargzTOC = false, false
 	}
 
 	if hasZstdChunkedTOC {
 		zstdChunkedTOCDigest, err := digest.Parse(zstdChunkedTOCDigestString)
-		if err != nil {
-			return nil, fmt.Errorf("parsing zstd:chunked TOC digest %q: %w", zstdChunkedTOCDigestString, err)
+		if err == nil {
+			return makeZstdChunkedDiffer(store, blobSize, zstdChunkedTOCDigest, annotations, iss, pullOptions)
```

Comment:
https://github.com/containers/storage/issues/2115#issuecomment-2379738952 talks about the case where a registry does not support ranged requests; that should be detected somewhere in `makeZstdChunkedDiffer` and turned into the “fall back” error. Similarly for the other differ.